Source/CarthageKit/VersionFile.swift

```diff
@@ -0,0 +1,272 @@
+/// Carthage/Build directory which associates its commitish with
+/// the MD5s of the built frameworks for each platform in order
+/// to allow those frameworks to be skipped in future builds.
+///
+/// Returns a signal that succeeds once the file has been created.
+public func createVersionFileForDependency(_ dependency: Dependency<PinnedVersion>, forPlatforms platforms: Set<Platform>, buildProductURLs: [URL], rootDirectoryURL: URL) -> SignalProducer<(), CarthageError> {
+	var platformCaches: [String: [CachedFramework]] = [:]
+
+	let platformsToCache = platforms.isEmpty ? Set(Platform.supportedPlatforms) : platforms
+	for platform in platformsToCache {
+		platformCaches[platform.rawValue] = []
+	}
+
+
+	let writeVersionFile = currentXcodeVersion()
+		.attemptMap { xcodeVersion -> Result<(), CarthageError> in
+			let rootBinariesURL = rootDirectoryURL.appendingPathComponent(CarthageBinariesFolderPath, isDirectory: true).resolvingSymlinksInPath()
+			let versionFileURL = rootBinariesURL.appendingPathComponent(".\(dependency.project.name).version")
+
+			let versionFile = VersionFile(
+				commitish: dependency.version.commitish,
+				xcodeVersion: xcodeVersion,
+				macOS: platformCaches[Platform.macOS.rawValue],
+				iOS: platformCaches[Platform.iOS.rawValue],
+				watchOS: platformCaches[Platform.watchOS.rawValue],
+				tvOS: platformCaches[Platform.tvOS.rawValue])
+
+			return versionFile.write(to: versionFileURL)
+	}
+
+	if !buildProductURLs.isEmpty {
+		return SignalProducer<URL, CarthageError>(buildProductURLs)
+		.flatMap(.merge) { url -> SignalProducer<String, CarthageError> in
+			let platformName = url.deletingLastPathComponent().lastPathComponent
+			let frameworkName = url.deletingPathExtension().lastPathComponent
+			let frameworkURL = url.appendingPathComponent(frameworkName, isDirectory: false)
+			return md5ForFileAtURL(frameworkURL)
+				.on(value: { md5 in
+					let cachedFramework = CachedFramework(name: frameworkName, md5: md5)
+					if var frameworks = platformCaches[platformName] {
+						frameworks.append(cachedFramework)
+						platformCaches[platformName] = frameworks
+					}
+				})
+		}
+		.then(writeVersionFile)
+	} else {
+		// Write out an empty version file for dependencies with no built frameworks, so cache builds can differentiate between
+		// no cache and a dependency that has no frameworks
+		return writeVersionFile
+	}
+}
+
+/// Determines whether a dependency can be skipped because it is
+/// already cached.
+///
+/// If a set of platforms is not provided, all platforms are checked.
+///
+/// Returns an optional bool which is nil if no version file exists,
+/// otherwise true if the version file matches and the build can be
+/// skipped or false if there is a mismatch of some kind.
+public func versionFileMatchesDependency(_ dependency: Dependency<PinnedVersion>, forPlatforms platforms: Set<Platform>, rootDirectoryURL: URL) -> SignalProducer<Bool?, CarthageError> {
```

Comment:
This could use a Swift 3 name.

```swift
public func versionFileMatches(_ dependency: Dependency<PinnedVersion>, for platforms: Set<Platform>, rootDirectory: URL) -> SignalProducer<Bool?, CarthageError> {
```